2.9BSD/usr/man/cat1/Mail.1

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

NAME
     mail - send and receive mail

SYNOPSIS
     Mail [ -f [ name ] ] [ people ...  ]

INTRODUCTION
     _M_a_i_l is a intelligent mail processing system, which has a
     command syntax reminiscent of _e_d with lines replaced by mes-
     sages.

     _S_e_n_d_i_n_g _m_a_i_l.  To send a message to one or more other peo-
     ple, _m_a_i_l can be invoked with arguments which are the names
     of people to send to.  You are then expected to type in your
     message, followed by an EOT (control-D) at the beginning of
     a line.  The section below, labeled _R_e_p_l_y_i_n_g _t_o _o_r _o_r_i_g_i_n_a_t_-
     _i_n_g _m_a_i_l, describes some features of _m_a_i_l available to help
     you compose your letter.

     _R_e_a_d_i_n_g _m_a_i_l.  In normal usage, _m_a_i_l is given no arguments
     and checks your mail out of the post office, then printing
     out a one line header of each message there.  The current
     message is initially the first message (numbered 1) and can
     be printed using the print command (which can be abbreviated
     p).  You can move among the messages much as you move
     between lines in _e_d, with the commands `+' and `-' moving
     backwards and forwards, and simple numbers typing the
     addressed message.

     _D_i_s_p_o_s_i_n_g _o_f _m_a_i_l.  After examining a message you can delete
     (d) the message or reply (r) to it.  Deletion causes the
     _m_a_i_l program to forget about the message.  This is not
     irreversible, the message can be undeleted (u) by giving its
     number, or the _m_a_i_l session can be aborted by giving the
     exit (x) command.  Deleted messages will, however, usually
     disappear never to be seen again.

     _S_p_e_c_i_f_y_i_n_g _m_e_s_s_a_g_e_s.  Commands such as print and delete
     often can be given a list of message numbers as argument to
     apply to a number of messages at once.  Thus ``delete 1 2''
     deletes messages 1 and 2, while ``delete 1-5'' deletes mes-
     sages 1 through 5.  The special name ``*'' addresses all
     messages, and ``$'' addresses the last message; thus the
     command top which prints the first few lines of a message
     could be used in ``top *'' to print the first few lines of
     all messages.

     _R_e_p_l_y_i_n_g _t_o _o_r _o_r_i_g_i_n_a_t_i_n_g _m_a_i_l.  You can use the reply com-
     mand to set up a response to a message, sending it back to
     the person who it was from.  Text you then type in, up to an
     end-of-file (or a line consisting only of a ``.'') defines
     the contents of the message.  While you are composing a

Printed 5/16/83                                                 1

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

     message, _m_a_i_l treats lines beginning with the character `~'
     specially.  For instance, typing ``~m'' (alone on a line)
     will place a copy of the current message into the response
     right shifting it by a tabstop.  Other escapes will set up
     subject fields, add and delete recipients to the message and
     allow you to escape to an editor to revise the message or to
     a shell to run some commands.  (These options will be given
     in the summary below.)

     _E_n_d_i_n_g _a _m_a_i_l _p_r_o_c_e_s_s_i_n_g _s_e_s_s_i_o_n.  You can end a _m_a_i_l ses-
     sion with the quit (q) command.  Messages which have been
     examined go to your _m_b_o_x file unless they have been deleted
     or preserved in which case they are discarded.  Unexamined
     messages go back to the post office.  The -f option causes
     _m_a_i_l to read in the contents of your _m_b_o_x (or the specified
     file) for processing; when you quit _m_a_i_l writes undeleted
     messages back to this file.

     _P_e_r_s_o_n_a_l _a_n_d _s_y_s_t_e_m_w_i_d_e _d_i_s_t_r_i_b_u_t_i_o_n _l_i_s_t_s.  It is also pos-
     sible to create a personal distribution lists so that, for
     instance, you can send mail to ``cohorts'' and have it go to
     a group of people.  Such lists can be defined by placing a
     line like

          alias cohorts bill ozalp sklower jkf mark cory:kridle

     in the file .mailrc in your home directory.  The current
     list of such aliases can be displayed by the alias (a) com-
     mand in _m_a_i_l. System wide distribution lists can be created
     by editing /usr/lib/aliases, see _a_l_i_a_s_e_s(5) and _d_e_l_i_v_e_r_-
     _m_a_i_l(8); these are kept in a slightly different syntax.  In
     mail you send, personal aliases will be expanded in mail
     sent to others so that they will be able to reply to the
     recipients.  System wide _a_l_i_a_s_e_s are not expanded when the
     mail is sent, but any reply returned to the machine will
     have the system wide alias expanded as all mail goes through
     _d_e_l_i_v_e_r_m_a_i_l. If you edit /usr/lib/aliases, you must run the
     program _n_e_w_a_l_i_a_s_e_s(1).

     _N_e_t_w_o_r_k _m_a_i_l (_A_R_P_A, _U_U_C_P, _B_e_r_k_n_e_t)  Mail to sites on the
     ARPA network and sites using UUCP can be sent using
     ``name@site'' for ARPA-net sites or ``machine!user'' for
     UUCP sites, provided appropriate gateways are known to the
     system.  (Be sure to escape the ! in UUCP sites when giving
     it on a _c_s_h command line by preceding it with a \.  Machines
     on an instance of the Berkeley network are addressed as
     ``machine:user'', e.g. ``csvax:bill''.  When addressed from
     the arpa-net, ``csvax:bill'' is known as
     ``csvax.bill@berkeley''.

     _M_a_i_l has a number of options which can be set in the ._m_a_i_l_r_c
     file to alter its behavior; thus ``set askcc'' enables the

Printed 5/16/83                                                 2

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

     ``askcc'' feature.  (These options are summarized below.)

SUMMARY
     (Adapted from the `Mail Reference Manual') Each command is
     typed on a line by itself, and may take arguments following
     the command word.  The command need not be typed in its
     entirety - the first command which matches the typed prefix
     is used.  For the commands which take message lists as argu-
     ments, if no message list is given, then the next message
     forward which satisfies the command's requirements is used.
     If there are no messages forward of the current message, the
     search proceeds backwards, and if there are no good messages
     at all, _m_a_i_l types ``No applicable messages'' and aborts the
     command.

     -           Goes to the previous message and prints it out.
                 If given a numeric argument _n , goes to the _n _t_h
                 previous message and prints it.

     ?           Prints a brief summary of commands.

     !           Executes the UNIX shell command which follows.

     alias       (a) With no arguments, prints out all
                 currently-defined aliases.  With one argument,
                 prints out that alias.  With more than one argu-
                 ment, adds the users named in the second and
                 later arguments to the alias named in the first
                 argument.

     chdir       (c) Changes the user's working directory to that
                 specified, if given.  If no directory is given,
                 then changes to the user's login directory.

     delete      (d) Takes a list of messages as argument and
                 marks them all as deleted.  Deleted messages
                 will not be saved in _m_b_o_x , nor will they be
                 available for most other commands.

     dp          (also dt) Deletes the current message and prints
                 the next message.  If there is no next message,
                 _m_a_i_l says ``at EOF.''

     edit        (e) Takes a list of messages and points the text
                 editor at each one in turn.  On return from the
                 editor, the message is read back in.

     exit        (ex or x) Effects an immediate return to the
                 Shell without modifying the user's system mail-
                 box, his _m_b_o_x file, or his edit file in -f .

     from        (f) Takes a list of messages and prints their

Printed 5/16/83                                                 3

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

                 message headers.

     headers     (h) Lists the current range of headers, which is
                 an 18 message group.  If a ``+'' argument is
                 given, then the next 18 message group is
                 printed, and if a ``-'' argument is given, the
                 previous 18 message group is printed.

     help        A synonym for ?

     hold        (ho, also preserve) Takes a message list and
                 marks each message therein to be saved in the
                 user's system mailbox instead of in _m_b_o_x. Does
                 not override the delete command.

     mail        (m) Takes as argument login names and distribu-
                 tion group names and sends mail to those people.

     next        (n like + or CR) Goes to the next message in
                 sequence and types it.  With an argument list,
                 types the next matching message.

     preserve    A synonym for hold.

     print       (p) Takes a message list and types out each mes-
                 sage on the user's terminal.

     quit        (q) Terminates the session, saving all
                 undeleted, unsaved messages in the user's _m_b_o_x
                 file in his login directory, preserving all mes-
                 sages marked with hold or preserve or never
                 referenced in his system mailbox, and removing
                 all other messages from his system mailbox.  If
                 new mail has arrived during the session, the
                 message ``You have new mail'' is given.  If
                 given while editing a mailbox file with the -f
                 flag, then the edit file is rewritten.  A return
                 to the Shell is effected, unless the rewrite of
                 edit file fails, in which case the user can
                 escape with the exit command.

     reply       (r) Takes a message list and sends mail to each
                 message author just like the mail command.  The
                 default message must not be deleted.

     respond     A synonym for reply .

     save        (s) Takes a message list and a filename and
                 appends each message in turn to the end of the
                 file.  The filename in quotes, followed by the
                 line count and character count is echoed on the
                 user's terminal.

Printed 5/16/83                                                 4

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

     set         (se) With no arguments, prints all variable
                 values.  Otherwise, sets option.  Arguments are
                 of the form ``option=value'' or ``option.''

     shell       (sh) Invokes an interactive version of the
                 shell.

     size        Takes a message list and prints out the size in
                 characters of each message.

     top         Takes a message list and prints the top few
                 lines of each.  The number of lines printed is
                 controlled by the variable toplines and defaults
                 to five.

     type        (t) A synonym for print .

     unalias     Takes a list of names defined by alias commands
                 and discards the remembered groups of users.
                 The group names no longer have any significance.

     undelete    (u) Takes a message list and marks each one as
                 _n_o_t being deleted.

     unset       Takes a list of option names and discards their
                 remembered values; the inverse of set .

     visual      (v) Takes a message list and invokes the display
                 editor on each message.

     write       (w) A synonym for save .

     xit         (x) A synonym for exit .

     Here is a summary of the tilde escapes, which are used when
     composing messages to perform special functions.  Tilde
     escapes are only recognized at the beginning of lines.  The
     name ``tilde escape'' is somewhat of a misnomer since the
     actual escape character can be set by the option escape.

     ~!command   Execute the indicated shell command, then return
                 to the message.

     ~c name ... Add the given names to the list of carbon copy
                 recipients.

     ~d          Read the file ``dead.letter'' from your home
                 directory into the message.

     ~e          Invoke the text editor on the message collected
                 so far.  After the editing session is finished,
                 you may continue appending text to the message.

Printed 5/16/83                                                 5

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

     ~h          Edit the message header fields by typing each
                 one in turn and allowing the user to append text
                 to the end or modify the field by using the
                 current terminal erase and kill characters.

     ~m messages Read the named messages into the message being
                 sent, shifted right one tab.  If no messages are
                 specified, read the current message.

     ~p          Print out the message collected so far, prefaced
                 by the message header fields.

     ~q          Abort the message being sent, copying the mes-
                 sage to ``dead.letter'' in your home directory
                 if save is set.

     ~r filename Read the named file into the message.

     ~s string   Cause the named string to become the current
                 subject field.

     ~t name ... Add the given names to the direct recipient
                 list.

     ~v          Invoke an alternate editor (defined by the
                 VISUAL option) on the message collected so far.
                 Usually, the alternate editor will be a screen
                 editor.  After you quit the editor, you may
                 resume appending text to the end of your mes-
                 sage.

     ~w filename Write the message onto the named file.

     ~|command   Pipe the message through the command as a
                 filter.  If the command gives no output or ter-
                 minates abnormally, retain the original text of
                 the message.  The command _f_m_t(1) is often used
                 as _c_o_m_m_a_n_d to rejustify the message.

     ~~string    Insert the string of text in the message pre-
                 faced by a single ~.  If you have changed the
                 escape character, then you should double that
                 character in order to send it.

     Options are controlled via the set and unset commands.
     Options may be either binary, in which case it is only sig-
     nificant to see whether they are set or not, or string, in
     which case the actual value is of interest.  The binary
     options include the following:

     append         Causes messages saved in _m_b_o_x to be appended
                    to the end rather than prepended.  (This is

Printed 5/16/83                                                 6

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

                    set in /usr/lib/Mail.rc on version 7 sys-
                    tems.)

     ask            Causes _m_a_i_l to prompt you for the subject of
                    each message you send.  If you respond with
                    simply a newline, no subject field will be
                    sent.

     askcc          Causes you to be prompted for additional car-
                    bon copy recipients at the end of each mes-
                    sage.  Responding with a newline indicates
                    your satisfaction with the current list.

     autoprint      Causes the delete command to behave like dp -
                    thus, after deleting a message, the next one
                    will be typed automatically.

     ignore         Causes interrupt signals from your terminal
                    to be ignored and echoed as @'s.

     metoo          Usually, when a group is expanded that con-
                    tains the sender, the sender is removed from
                    the expansion.  Setting this option causes
                    the sender to be included in the group.

     quiet          Suppresses the printing of the version when
                    first invoked.

     save           Causes the message collected prior to a
                    interrupt to be saved on the file
                    ``dead.letter'' in your home directory on
                    receipt of two interrupts (or after a ~q.)

     The following options have string values:

     EDITOR         Pathname of the text editor to use in the
                    edit command and ~e escape.  If not defined,
                    then a default editor is used.

     SHELL          Pathname of the shell to use in the ! command
                    and the ~! escape.  A default shell is used
                    if this option is not defined.

     VISUAL         Pathname of the text editor to use in the
                    visual command and ~v escape.

     escape         If defined, the first character of this
                    option gives the character to use in the
                    place of ~ to denote escapes.

     record         If defined, gives the pathname of the file
                    used to record all outgoing mail.  If not

Printed 5/16/83                                                 7

MAIL(1)             UNIX Programmer's Manual              MAIL(1)

                    defined, then outgoing mail is not so saved.

     toplines       If defined, gives the number of lines of a
                    message to be printed out with the top com-
                    mand; normally, the first five lines are
                    printed.

FILES
     /usr/spool/mail/*        post office
     ~/mbox                   your old mail
     ~/.mailrc                file giving initial mail commands
     /tmp/R#                  temporary for editor escape
     /usr/lib/Mail.help*      help files
     /usr/lib/Mail.rc         system initialization file
     /bin/mail                to do actual mailing
     /etc/delivermail         postman

SEE ALSO
     fmt(1), mail(1), newaliases(1), aliases(5), delivermail(8)
     _T_h_e _M_a_i_l _R_e_f_e_r_e_n_c_e _M_a_n_u_a_l by Kurt Shoens

AUTHOR
     Kurt Shoens

BUGS
     Doesn't know how to handle standard addresses such as
     user@machine.UUCP.
     It isn't MH.

Printed 5/16/83                                                 8



2.9BSD/usr/src/cmd/mail.c
#include <stdio.h>
#include <pwd.h>
#include <utmp.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <setjmp.h>
#include <whoami.h>

#define	MSGS		/* pipe mail to "msgs" to /usr/ucb/msgs */
/*copylet flags */
	/*remote mail, add rmtmsg */
#define REMOTE	1
	/* zap header and trailing empty line */
#define ZAP	3
#define ORDINARY 2
#define	FORWARD	4
#define	LSIZE	256
#define	MAXLET	300	/* maximum number of letters */
#define	MAILMODE (~0600)		/* mode of created mail */

char	line[LSIZE];
char	resp[LSIZE];
struct let {
	long	adr;
	char	change;
} let[MAXLET];
int	nlet	= 0;
char	lfil[50];
long	iop, time();
char	lettmp[] = "/tmp/maXXXXX";
char	maildir[] = "/usr/spool/mail/";
char	mailfile[] = "/usr/spool/mail/xxxxxxxxxxxxxxxxxxxxxxx";
char	dead[] = "dead.letter";
char	forwmsg[] = " forwarded\n";
char	*curlock;
int	lockerror;
FILE	*tmpf;
FILE	*malf;
char	*my_name;
struct	passwd	*getpwuid();
int	error;
int	locked;
int	changed;
int	forward;
char	from[] = "From ";
long	ftell();
int	delete();
char	*ctime();
int	flgf;
int	flgp;
int	delflg = 1;
jmp_buf	sjbuf;
char	hostname[32];

main(argc, argv)
char **argv;
{
	register i;
	char sobuf[BUFSIZ];

	setbuf(stdout, sobuf);
	mktemp(lettmp);
	unlink(lettmp);
	if (my_name == NULL) {
		struct passwd *pwent;
		pwent = getpwuid(getuid());
		if (pwent==NULL)
			my_name = "???";
		else
			my_name = pwent->pw_name;
	}
	if(setjmp(sjbuf)) done();
	for (i = SIGHUP; i <= SIGTERM; i++)
		setsig(i, delete);
	tmpf = fopen(lettmp, "w");
	if (tmpf == NULL) {
		fprintf(stderr, "mail: cannot open %s for writing\n", lettmp);
		done();
	}
	if (argv[0][0] != 'r' &&	/* no favors for rmail*/
	   (argc == 1 || argv[1][0] == '-'))
		printmail(argc, argv);
	else
		sendmail(argc, argv);
	done();
}

setsig(i, f)
int i;
int (*f)();
{
	if(signal(i, SIG_IGN) != SIG_IGN)
		signal(i, f);
}

printmail(argc, argv)
char **argv;
{
	int flg, i, j, print;
	char *p, *getarg();

	setuid(getuid());
	cat(mailfile, maildir, my_name);
	for (; argc>1; argv++, argc--) {
		if (argv[1][0]=='-') {
			if (argv[1][1]=='q')
				delflg = 0;
			else if (argv[1][1]=='p') {
				flgp++;
				delflg = 0;
			} else if (argv[1][1]=='f') {
				if (argc>=3) {
					strcpy(mailfile, argv[2]);
					argv++;
					argc--;
				}
			} else if (argv[1][1]=='r') {
				forward = 1;
			} else {
				fprintf(stderr, "mail: unknown option %c\n", argv[1][1]);
				done();
			}
		} else
			break;
	}
	malf = fopen(mailfile, "r");
	if (malf == NULL) {
		fprintf(stdout, "No mail.\n");
		return;
	}
	lock(mailfile);
	copymt(malf, tmpf);
	fclose(malf);
	fclose(tmpf);
	unlock();
	tmpf = fopen(lettmp, "r");

	changed = 0;
	print = 1;
	for (i = 0; i < nlet; ) {
		j = forward ? i : nlet - i - 1;
		if(setjmp(sjbuf)) {
			print=0;
		} else {
			if (print)
				copylet(j, stdout, ORDINARY);
			print = 1;
		}
		if (flgp) {
			i++;
			continue;
		}
		setjmp(sjbuf);
		fprintf(stdout, "? ");
		fflush(stdout);
		if (fgets(resp, LSIZE, stdin) == NULL)
			break;
		switch (resp[0]) {

		default:
			fprintf(stderr, "usage\n");
		case '?':
			print = 0;
			fprintf(stderr, "q\tquit\n");
			fprintf(stderr, "x\texit without changing mail\n");
			fprintf(stderr, "p\tprint\n");
			fprintf(stderr, "s[file]\tsave (default mbox)\n");
			fprintf(stderr, "w[file]\tsame without header\n");
			fprintf(stderr, "-\tprint previous\n");
			fprintf(stderr, "d\tdelete\n");
			fprintf(stderr, "+\tnext (no delete)\n");
			fprintf(stderr, "m user\tmail to user\n");
			fprintf(stderr, "! cmd\texecute cmd\n");
			break;

		case '+':
		case 'n':
		case '\n':
			i++;
			break;
		case 'x':
			changed = 0;
		case 'q':
			goto donep;
		case 'p':
			break;
		case '^':
		case '-':
			if (--i < 0)
				i = 0;
			break;
		case 'y':
		case 'w':
		case 's':
			flg = 0;
			if (resp[1] != '\n' && resp[1] != ' ') {
				printf("illegal\n");
				flg++;
				print = 0;
				continue;
			}
			if (resp[1] == '\n' || resp[1] == '\0')
				cat(resp+1, "mbox", "");
			for (p = resp+1; (p = getarg(lfil, p)) != NULL; ) {
				malf = fopen(lfil, "a");
				if (malf == NULL) {
					fprintf(stdout, "mail: cannot append to %s\n", lfil);
					flg++;
					continue;
				}
				copylet(j, malf, resp[0]=='w'? ZAP: ORDINARY);
				fclose(malf);
			}
			if (flg)
				print = 0;
			else {
				let[j].change = 'd';
				changed++;
				i++;
			}
			break;
		case 'm':
			flg = 0;
			if (resp[1] == '\n' || resp[1] == '\0') {
				i++;
				continue;
			}
			if (resp[1] != ' ') {
				printf("invalid command\n");
				flg++;
				print = 0;
				continue;
			}
			for (p = resp+1; (p = getarg(lfil, p)) != NULL; )
				if (!sendrmt(j, lfil))	/* couldn't send it */
					flg++;
			if (flg)
				print = 0;
			else {
				let[j].change = 'd';
				changed++;
				i++;
			}
			break;
		case '!':
			system(resp+1);
			printf("!\n");
			print = 0;
			break;
		case 'd':
			let[j].change = 'd';
			changed++;
			i++;
			if (resp[1] == 'q')
				goto donep;
			break;
		}
	}
   donep:
	if (changed)
		copyback();
}

copyback()	/* copy temp or whatever back to /usr/spool/mail */
{
	register i, n, c;
	int new = 0;
	struct stat stbuf;

	signal(SIGINT, SIG_IGN);
	signal(SIGHUP, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	lock(mailfile);
	stat(mailfile, &stbuf);
	if (stbuf.st_size != let[nlet].adr) {	/* new mail has arrived */
		malf = fopen(mailfile, "r");
		if (malf == NULL) {
			fprintf(stdout, "mail: can't re-read %s\n", mailfile);
			done();
		}
		fseek(malf, let[nlet].adr, 0);
		fclose(tmpf);
		tmpf = fopen(lettmp, "a");
		fseek(tmpf, let[nlet].adr, 0);
		while ((c = fgetc(malf)) != EOF)
			fputc(c, tmpf);
		fclose(malf);
		fclose(tmpf);
		tmpf = fopen(lettmp, "r");
		let[++nlet].adr = stbuf.st_size;
		new = 1;
	}
	malf = fopen(mailfile, "w");
	if (malf == NULL) {
		fprintf(stderr, "mail:  can't rewrite %s\n", lfil);
		done();
	}
	n = 0;
	for (i = 0; i < nlet; i++)
		if (let[i].change != 'd') {
			copylet(i, malf, ORDINARY);
			n++;
		}
	fclose(malf);
	if (new)
		fprintf(stdout, "new mail arrived\n");
	unlock();
}

copymt(f1, f2)	/* copy mail (f1) to temp (f2) */
FILE *f1, *f2;
{
	long nextadr;

	nlet = nextadr = 0;
	let[0].adr = 0;
	while (fgets(line, LSIZE, f1) != NULL) {
		if (isfrom(line))
			let[nlet++].adr = nextadr;
		nextadr += strlen(line);
		fputs(line, f2);
	}
	let[nlet].adr = nextadr;	/* last plus 1 */
}

copylet(n, f, type) FILE *f;
{	int ch;
	long k;
	fseek(tmpf, let[n].adr, 0);
	k = let[n+1].adr - let[n].adr;
	while(k-- > 1L && (ch=fgetc(tmpf))!='\n')
		if(type!=ZAP) fputc(ch,f);
	if(type==REMOTE) {
		gethostname(hostname, sizeof (hostname));
		fprintf(f, " remote from %s\n", hostname);
	}
	else if (type==FORWARD)
		fprintf(f, forwmsg);
	else if(type==ORDINARY)
		fputc(ch,f);
	while(k-->1L)
		fputc(ch=fgetc(tmpf), f);
	if(type!=ZAP || ch!= '\n')
		fputc(fgetc(tmpf), f);
}

isfrom(lp)
register char *lp;
{
	register char *p;

	for (p = from; *p; )
		if (*lp++ != *p++)
			return(0);
	return(1);
}

sendmail(argc, argv)
char **argv;
{

	time(&iop);
	fprintf(tmpf, "%s%s %s", from, my_name, ctime(&iop));
	iop = ftell(tmpf);
	flgf = 1;
	while (fgets(line, LSIZE, stdin) != NULL) {
		if (line[0] == '.' && line[1] == '\n')
			break;
		if (isfrom(line))
			fputs(">", tmpf);
		fputs(line, tmpf);
		flgf = 0;
	}
	fputs("\n", tmpf);
	nlet = 1;
	let[0].adr = 0;
	let[1].adr = ftell(tmpf);
	fclose(tmpf);
	if (flgf)
		return;
	tmpf = fopen(lettmp, "r");
	if (tmpf == NULL) {
		fprintf(stderr, "mail:  cannot reopen %s for reading\n", lettmp);
		return;
	}
	while (--argc > 0)
		if (!send(0, *++argv))	/* couldn't send to him */
			error++;
	if (error && safefile(dead)) {
		setuid(getuid());
		malf = fopen(dead, "w");
		if (malf == NULL) {
			fprintf(stdout, "mail:  cannot open %s\n", dead);
			fclose(tmpf);
			return;
		}
		copylet(0, malf, ZAP);
		fclose(malf);
		fprintf(stdout, "Mail saved in %s\n", dead);
	}
	fclose(tmpf);
}

sendrmt(n, name)
char *name;
{
	FILE *rmf, *popen();
	register char *p;
	char rsys[64], cmd[64];
	register local, pid;
	int ret, sts;

	local = 0;
	if (*name=='!')
		name++;
	for(p=rsys; *name!='!'; *p++ = *name++)
		if (*name=='\0') {
			local++;
			break;
		}
	*p = '\0';
	if ((!local && *name=='\0') || (local && *rsys=='\0')) {
		fprintf(stdout, "null name\n");
		return(0);
	}
	if ((pid = fork()) == -1) {
		fprintf(stderr, "mail:  can't create proc for remote\n");
		return(0);
	}
	if (pid) {
		while ((ret = wait(&sts)) != pid) {
			if (ret == -1)
				return(0);
		}
		return(!sts);
	}
	setuid(getuid());
	if (local)
		sprintf(cmd, "mail %s", rsys);
	else {
		if (index(name+1, '!'))
			sprintf(cmd, "uux - %s!rmail \\(%s\\)", rsys, name+1);
		else
			sprintf(cmd, "uux - %s!rmail %s", rsys, name+1);
	}
	if ((rmf=popen(cmd, "w")) == NULL)
		exit(1);
	copylet(n, rmf, local?  FORWARD : REMOTE);
	exit(pclose(rmf) != 0);
}

send(n, name)	/* send letter n to name */
int n;
char *name;
{
	char file[50];
	register char *p;
	register mask;
	struct passwd *pw, *getpwnam();
	struct stat statb;
#ifdef MSGS
	FILE *pip;
	char line[256];
	int count;
#endif

	for (p = name; *p != '!' && *p != '^' && *p != '\0'; p++)
		;
	if (*p == '!' || *p == '^')
		return(sendrmt(n, name));
#ifdef MSGS
/*
 * modification to call Berkeley 'msgs' program when the 'recipient'
 * is "msgs".  Assumption is that 'n' is 0, no REMOTE, FORWARD, etc.
 * pipe the letter to "msgs -s"  -- pag 11/10/79
 */
	if(!strcmp(name,"msgs"))	/* this letter is really a mesg */
	{
	    pip = popen("/usr/ucb/msgs -s", "w");
	    if (pip == NULL)
		return(0);
	    fseek(tmpf,0L,0);
	    count = 0;
	    while(fgets(line,256,tmpf) != NULL)
	    {
		count++;
		if(count == 2)
		{
		    fprintf(pip,"To: msgs\n");
 /*
  * check 2nd line for a Berkeley 'Mail'-added "To:" line....
  * we don't want to add 2 "To:"'s
  */
		    if(!strncmp(line,"To: msgs",8))
			continue;
		}
		fputs(line,pip);
	    }
	    pclose(pip);
	    return(1);
	}
#endif
	if ((pw = getpwnam(name)) == NULL) {
		fprintf(stdout, "mail:  can't send to %s\n", name);
		return(0);
	}
	cat(file, maildir, name);
	if (stat(file, &statb) >= 0 && (statb.st_mode & S_IFMT) == S_IFDIR) {
		strcat(file, "/");
		strcat(file, name);
	}
	mask = umask(MAILMODE);
	if (!safefile(file))
		return(0);
	lock(file);
	malf = fopen(file, "a");
	umask(mask);
	if (malf == NULL) {
		unlock();
		fprintf(stdout, "mail:  cannot append to %s\n", file);
		return(0);
	}
	chown(file, pw->pw_uid, pw->pw_gid);
	copylet(n, malf, ORDINARY);
	fclose(malf);
	unlock();
	return(1);
}

delete(i)
{
	setsig(i, delete);
	fprintf(stderr, "\n");
	if(delflg)
		longjmp(sjbuf, 1);
	done();
}

done()
{
	if(!lockerror)
		unlock();
	unlink(lettmp);
	exit(error+lockerror);
}

lock(file)
char *file;
{
	struct stat stbuf;

	if (locked || flgf)
		return;
	if (stat(file, &stbuf)<0)
		return;
	if (stbuf.st_mode&01) { 	/* user x bit is the lock */
		if (stbuf.st_ctime+60 >= time((long *)0)) {
			fprintf(stderr, "%s busy; try again in a minute\n", file);
			lockerror++;
			done();
		}
	}
	locked = stbuf.st_mode & ~01;
	curlock = file;
	chmod(file, stbuf.st_mode|01);
}

unlock()
{
	if (locked)
		chmod(curlock, locked);
	locked = 0;
}

cat(to, from1, from2)
char *to, *from1, *from2;
{
	int i, j;

	j = 0;
	for (i=0; from1[i]; i++)
		to[j++] = from1[i];
	for (i=0; from2[i]; i++)
		to[j++] = from2[i];
	to[j] = 0;
}

char *getarg(s, p)	/* copy p... into s, update p */
register char *s, *p;
{
	while (*p == ' ' || *p == '\t')
		p++;
	if (*p == '\n' || *p == '\0')
		return(NULL);
	while (*p != ' ' && *p != '\t' && *p != '\n' && *p != '\0')
		*s++ = *p++;
	*s = '\0';
	return(p);
}

safefile(f)
	char *f;
{
	struct stat statb;

#ifdef	UCB_SYMLINKS
	if (lstat(f, &statb) < 0)
#else
	if (stat(f, &statb) < 0)
#endif
		return(1);
	if (statb.st_nlink != 1 || (statb.st_mode & S_IFMT) == S_IFLNK) {
		fprintf(stderr, "mail:  %s has more than one link or is a symbolic link\n", f);
		return(0);
	}
	return(1);
}
